### PR TITLE
 [Enhancement]Support mixed-sign ramp indices in LegalizeNegativeIndex

### PR DESCRIPTION
**Pull request by TerminusAkivili**
## Summary

- Extend `LegalizeNegativeIndex` to handle constant `Ramp` indices with mixed negative and non-negative lanes.
- Rewrite only lanes proven negative, preserving non-negative lanes.
- Conservatively widen affected block read/write regions when the original region starts below zero.

## Motivation

TileLang already legalizes Python-style negative indices in scalar and all-negative vector cases:

```python
A[-1]
A[T.Ramp(-4, 1, 4)]
```

This PR extends the same behavior to constant mixed-sign ramps:

```python
A[T.Ramp(-2, 1, 4)]  # lanes: [-2, -1, 0, 1]
```

The current pass classifies the whole ramp as unknown-sign and leaves it unchanged, even though each lane is constant and can be legalized independently.

## Kernel

This minimal circular-buffer example exercises a vectorized load whose constant ramp crosses the zero boundary.

```python
import torch
import tilelang
from tilelang import tvm as tvm
import tilelang.language as T


CACHE_LEN = 1024
WINDOW = 4
WRAP_LEFT = 2


@T.prim_func
def sliding_window_kv_cache_unwrap(
    K_cache: T.Tensor((CACHE_LEN,), T.float32),
    K_window: T.Tensor((WINDOW,), T.float32),
):
    """Gather a KV window that crosses the circular-cache boundary.

    Logical window:
      K_cache[-2], K_cache[-1], K_cache[0], K_cache[1]

    Expected physical slots:
      K_cache[1022], K_cache[1023], K_cache[0], K_cache[1]
    """
    with T.Kernel(1, threads=1) as _:
        logical_window = T.Ramp(-WRAP_LEFT, 1, WINDOW)
        K_window[T.Ramp(0, 1, WINDOW)] = K_cache[logical_window]


def show_lowered_ir():
    mod = tvm.IRModule.from_expr(
        sliding_window_kv_cache_unwrap.with_attr("global_symbol", "main")
    )
    lowered = tilelang.transform.LegalizeNegativeIndex()(mod)["main"]
    print(lowered.script())


def run_kernel():
    backing = torch.empty(CACHE_LEN + WRAP_LEFT, device="cuda", dtype=torch.float32)
    backing[0] = -2000.0
    backing[1] = -1000.0
    backing[WRAP_LEFT:] = torch.arange(CACHE_LEN, device="cuda", dtype=torch.float32)

    k_cache_view = backing[WRAP_LEFT:]
    k_window = torch.empty(WINDOW, device="cuda", dtype=torch.float32)
    expected = torch.tensor([1022.0, 1023.0, 0.0, 1.0], device="cuda")

    kernel = tilelang.compile(
        sliding_window_kv_cache_unwrap,
        target="cuda",
        execution_backend="nvrtc",
    )
    kernel(k_cache_view, k_window)
    torch.cuda.synchronize()

    print("actual  :", k_window.detach().cpu().tolist())
    print("expected:", expected.detach().cpu().tolist())
    assert torch.allclose(k_window, expected)


if __name__ == "__main__":
    show_lowered_ir()
    if torch.cuda.is_available():
        run_kernel()
```

Before this patch, the mixed-sign access remains unlegalized:

```python
T.reads(K_cache[-2:2])
K_window[0:4] = K_cache[logical_window]
```

After this patch, the transform rewrites the lanes independently:

```python
T.reads(K_cache[0:1024])
K_window[0:4] = K_cache[T.Shuffle([1022, 1023, 0, 1], [0, 1, 2, 3])]
```

## Implementation

For unknown-sign constant ramps, the rewriter now checks each lane:

- negative lane: rewrite to `buffer_extent + lane`
- non-negative lane: keep unchanged
- unknown lane: leave the original expression unchanged

The rewritten vector is emitted as `T.Shuffle`, preserving per-lane semantics while leaving truly dynamic unknown-sign cases untouched. Existing scalar-negative and all-negative vector paths are unchanged.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved legalization of mixed-sign vector indices for loads and stores, ensuring correct handling in vectorized operations and preserving integer-width of buffer extents.
  * Block-level buffer regions with provably negative axis minima are conservatively expanded to full regions to prevent incorrect accesses.

* **Tests**
  * Added tests covering mixed-sign vectorized load/store legalization, including cases inside kernels and int64 extents.

[Review Change Stack](https://app.coderabbit.ai/change-stack/tile-ai/tilelang/pull/2225?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by LeiWang1999 — APPROVED** [review]

**Comment by LeiWang1999** [comment]
Thanks!
